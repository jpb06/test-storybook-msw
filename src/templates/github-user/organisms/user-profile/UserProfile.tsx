import CampaignIcon from '@mui/icons-material/Campaign';
import CommitIcon from '@mui/icons-material/Commit';
import LocationOnIcon from '@mui/icons-material/LocationOn';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import { CardContent, List } from '@mui/material';
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import IconButton from '@mui/material/IconButton';

import { GithubProfileQueryResult, useGithubProfileQuery } from '@api';
import { getRuntimeConfig } from '@logic/getRuntimeConfig';
import { TopAlert } from '@molecules';

import { ProfileInfo } from './molecules/profile-info/ProfileInfo';
import { CardSkeleton } from './molecules/skeletons/CardSkeleton';
import { UserAvatar } from './molecules/user-avatar/UserAvatar';

export const UserProfile = () => {
  const { isLoading, isError, data } = useGithubProfileQuery();
  const {
    public: { basePath },
  } = getRuntimeConfig();

  if (isLoading) {
    return <CardSkeleton />;
  }

  if (isError) {
    return (
      <TopAlert
        severity="error"
        errorText="We've been unable to load your profile"
      >
        <CardSkeleton />
      </TopAlert>
    );
  }

  const { followers, location, public_repos, name, email } =
    data as GithubProfileQueryResult;

  return (
    <Card>
      <CardHeader
        avatar={<UserAvatar name={name} />}
        action={
          <IconButton aria-label="settings">
            <MoreVertIcon />
          </IconButton>
        }
        title={name}
        subheader={email}
      />
      <CardMedia
        component="img"
        height="160"
        image={`${basePath}/img/github-logo.jpeg`}
        alt={name}
      />
      <CardContent>
        <List dense>
          <ProfileInfo Icon={LocationOnIcon} type="Location" value={location} />
          <ProfileInfo
            Icon={CommitIcon}
            type="Public repos"
            value={public_repos}
          />
          <ProfileInfo Icon={CampaignIcon} type="Followers" value={followers} />
        </List>
      </CardContent>
    </Card>
  );
};
