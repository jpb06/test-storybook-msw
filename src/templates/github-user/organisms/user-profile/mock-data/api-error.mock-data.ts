export const apiErrorMockData = {
  code: 500,
  message: 'Oh no!',
};
