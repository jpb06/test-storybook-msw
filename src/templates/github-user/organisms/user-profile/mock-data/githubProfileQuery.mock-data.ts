export const githubProfileQueryMockData = {
  info: {
    nickname: 'coolbro',
    email: 'cool.bro@yolo.org',
    name: 'Yolo McBro',
  },
  extra: {
    raw_info: {
      followers: 243,
      following: 345,
      public_repos: 12,
      public_gists: 1,
      location: 'Paris, France',
    },
  },
};
