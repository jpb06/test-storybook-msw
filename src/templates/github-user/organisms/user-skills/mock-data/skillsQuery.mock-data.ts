export const skillsQueryMockData = [
  {
    id: 1,
    name: 'Soft skills',
    skills: [
      { id: 6, name: 'Communication', idRoles: [3] },
      { id: 8, name: 'Information sharing', idRoles: [3] },
    ],
  },
  {
    id: 2,
    name: 'Management',
    skills: [
      {
        id: 7,
        name: 'Project drive',
        idRoles: [4],
      },
      { id: 9, name: 'Reporting', idRoles: [4] },
      { id: 11, name: 'Roadmap definition', idRoles: [4] },
    ],
  },
  {
    id: 3,
    name: 'Tech',
    skills: [
      { id: 1, name: 'jest', idRoles: [1, 4] },
      { id: 4, name: 'react', idRoles: [1, 4] },
      { id: 5, name: 'Typescript', idRoles: [1, 4] },
      { id: 10, name: 'Github actions', idRoles: [2] },
    ],
  },
];
