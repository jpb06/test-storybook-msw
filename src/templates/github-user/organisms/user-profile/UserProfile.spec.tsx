import { screen } from '@testing-library/react';

import { githubProfileQueryHandler } from '@api/msw-handlers';
import { render } from '@tests/renders/render';

import { UserProfile } from './UserProfile';
import { githubProfileQueryMockData } from './mock-data/githubProfileQuery.mock-data';

jest.mock('@logic/delay');

describe('UserProfile component', () => {
  describe('nominal case', () => {
    it('should match snapshot', async () => {
      githubProfileQueryHandler(githubProfileQueryMockData, 200, true);

      const { baseElement } = render(<UserProfile />);

      await screen.findByRole('img', { name: /yolo mcbro/i });

      expect(baseElement).toMatchSnapshot();
    });

    it('should display skills', async () => {
      githubProfileQueryHandler(githubProfileQueryMockData, 200, true);

      render(<UserProfile />);

      await screen.findByText(githubProfileQueryMockData.info.name);
      screen.getByText(githubProfileQueryMockData.info.email);
      screen.getByText(githubProfileQueryMockData.extra.raw_info.followers);
      screen.getByText(githubProfileQueryMockData.extra.raw_info.public_repos);
    });
  });

  describe('error case', () => {
    it('should match snapshot', async () => {
      githubProfileQueryHandler({ code: 500, message: 'Oh no!' }, 500, true);

      const { baseElement } = render(<UserProfile />);

      await screen.findByRole('alert');

      expect(baseElement).toMatchSnapshot();
    });

    it('should display an error when request failed', async () => {
      githubProfileQueryHandler({ code: 500, message: 'Oh no!' }, 500, true);

      render(<UserProfile />);

      await screen.findByText(/we've been unable to load your profile/i);
      expect(screen.getByTestId('ErrorOutlineIcon')).toBeInTheDocument();
    });
  });
});
