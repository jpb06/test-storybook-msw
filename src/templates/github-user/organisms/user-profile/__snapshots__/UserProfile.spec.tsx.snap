// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile component error case should match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-outlinedError MuiAlert-outlined css-1yc4i3e-MuiPaper-root-MuiAlert-root"
        role="alert"
      >
        <div
          class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="ErrorOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-acap47-MuiAlert-message"
        >
          We've been unable to load your profile
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-wave css-twt492-MuiSkeleton-root"
              style="width: 55px; height: 55px;"
            />
          </div>
          <div
            class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                style="width: 80%; height: 16px; margin-top: 14px; margin-bottom: 6px;"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                style="width: 40%; height: 14px; margin-bottom: 22px;"
              />
            </span>
          </div>
          <div
            class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
          >
            <button
              aria-label="settings"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MoreVertIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <img
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-mygpk5-MuiCardMedia-root"
          height="160"
          src="/front/img/github-logo.jpeg"
        />
        <div
          class="MuiCardContent-root css-10j7fi-MuiCardContent-root"
        >
          <ul
            class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
          >
            <li
              class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1yfhx2g-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-wave css-twt492-MuiSkeleton-root"
                  style="width: 40px; height: 40px;"
                />
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 20%; height: 14px;"
                  />
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 60%; height: 14px;"
                  />
                </p>
              </div>
            </li>
            <li
              class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1yfhx2g-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-wave css-twt492-MuiSkeleton-root"
                  style="width: 40px; height: 40px;"
                />
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 20%; height: 14px;"
                  />
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 60%; height: 14px;"
                  />
                </p>
              </div>
            </li>
            <li
              class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1yfhx2g-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-wave css-twt492-MuiSkeleton-root"
                  style="width: 40px; height: 40px;"
                />
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 20%; height: 14px;"
                  />
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1dk1h2u-MuiSkeleton-root"
                    style="width: 60%; height: 14px;"
                  />
                </p>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`UserProfile component nominal case should match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1ewlecp-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18949cl-MuiAvatar-root"
              >
                YM
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            Yolo McBro
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            cool.bro@yolo.org
          </span>
        </div>
        <div
          class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
        >
          <button
            aria-label="settings"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MoreVertIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <img
        alt="Yolo McBro"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="160"
        src="/front/img/github-logo.jpeg"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <ul
          class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-ax6k3r-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="LocationOnIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                Location
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                Paris, France
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-ax6k3r-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CommitIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.9 11c-.46-2.28-2.48-4-4.9-4s-4.44 1.72-4.9 4H2v2h5.1c.46 2.28 2.48 4 4.9 4s4.44-1.72 4.9-4H22v-2h-5.1zM12 15c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                Public repos
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                12
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-ax6k3r-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CampaignIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18 11v2h4v-2h-4zm-2 6.61c.96.71 2.21 1.65 3.2 2.39.4-.53.8-1.07 1.2-1.6-.99-.74-2.24-1.68-3.2-2.4-.4.54-.8 1.08-1.2 1.61zM20.4 5.6c-.4-.53-.8-1.07-1.2-1.6-.99.74-2.24 1.68-3.2 2.4.4.53.8 1.07 1.2 1.6.96-.72 2.21-1.65 3.2-2.4zM4 9c-1.1 0-2 .9-2 2v2c0 1.1.9 2 2 2h1v4h2v-4h1l5 3V6L8 9H4zm11.5 3c0-1.33-.58-2.53-1.5-3.35v6.69c.92-.81 1.5-2.01 1.5-3.34z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                Followers
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                243
              </p>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
